### PR TITLE
Integrate Mars DataFrame with Ray Dataset

### LINKED ISSUE #2229: Ray Dataset Integration
Support mars dataframe interactivity with ray dataset.
> Ray Datasets are the standard way to load and exchange data in Ray libraries and applications. Datasets provide basic distributed data transformations such as map, filter, and repartition, and are compatible with a variety of file formats, datasources, and distributed frameworks

It would be great if we can support convert mars dataframe to/from ray dataset, in this way mars can interact with all libs and frameworks running on ray, such as [xgboost on ray](https://github.com/ray-project/xgboost_ray), [raysgd](https://docs.ray.io/en/master/raysgd/raysgd.html)

**Describe the solution you'd like**
```python
# Convert mars dataframe to ray dataset
df: md.Dataframe = xxx
ds: ray.data.Dataset = df.to_ray_dataset()
ds: ray.data.Dataset = ray.data.from_mars(df)

# Use xgboost on ray to train from mars dataframe
mat = RayDMatrix(ds,
                 "label",  # # Will select this column as the label
                 distributed=True)

# Create mars dataframe from ray dataset
ds: ray.data.Dataset = xxx
df: md.Dataframe = ds.to_mars()
```

**Related issues**
#2112

### PR DESCRIPTION
**Pull request by vcfgv**
## What do these changes do?

In order that Mars can support converting Mars DataFrame to Ray Dataset, the following changes were made.

- Introduced a new Op named `read_raydataset` to help convert ObjectRefs in Ray Dataset into a Mars DataFrame while each as a chunk.
- Fetched ObjectRefs is sufficient for conversion from Mars DataFrame to Ray Dataset.

Furthermore, `xgboost_ray` is initially tested working with Mars DataFrame. 

**This PR is a clean slate replacing PR 2372.**

## Related issue number

Fixes #2229

### CODE REVIEW

File: mars/dataframe/contrib/raydataset/tests/test_mldataset.py [hunk 1]
```diff
@@ -90,8 +86,6 @@ async def test_convert_to_ray_mldataset(ray_large_cluster, create_cluster, test_
 
 @require_ray
 @pytest.mark.asyncio
-@pytest.mark.skipif(sklearn is None, reason='sklearn not installed')
-@pytest.mark.skipif(xgboost_ray is None, reason='xgboost_ray not installed')
```
Comment (wjsi) [inline mars/dataframe/contrib/raydataset/tests/test_mldataset.py:94]: Better to put this line back.

File: .github/workflows/platform-ci.yml [hunk 2]
```diff
@@ -90,6 +90,7 @@ jobs:
             if [ -n "$WITH_RAY" ]; then
               # Change back to `pip install ray[default]` when ray-1.6.0 is released.
               pip install https://s3-us-west-2.amazonaws.com/ray-wheels/latest/ray-2.0.0.dev0-cp38-cp38-manylinux2014_x86_64.whl
+              pip install xgboost_ray
```
[the remark this answers is not in the document]
Comment (chaokunyang) [reply .github/workflows/platform-ci.yml]: ray 1.5 and 1.6 doesn't allow actor name with `/` embedded, we may need to wait until 1.7 got released.

File: mars/dataframe/datasource/read_raydataset.py [hunk 3]
```diff
@@ -0,0 +1,115 @@
...
+        df = ray.get(refs)
+        ctx[out.key] = df
+
+    def __call__(self, index_value=None, columns_value=None, dtypes=None):
+        shape = (np.nan, len(dtypes))
+        return self.new_dataframe(None, shape, dtypes=dtypes, index_value=index_value,
+                                  columns_value=columns_value)
+
+
+def read_raydataset(refs, columns=None,
```
Comment (chaokunyang) [inline mars/dataframe/datasource/read_raydataset.py:101]: Can we also support pass ray.data.dataset.Dataset?

File: mars/dataframe/datasource/read_raydataset.py [hunk 4]
```diff
@@ -0,0 +1,115 @@
...
+    @property
+    def columns(self):
+        return self._columns
+
+    @property
+    def incremental_index(self):
+        return self._incremental_index
+
+    @classmethod
+    def _tile_partitioned(cls, op: "DataFrameReadRayDataset"):
```
Comment (chaokunyang) [inline mars/dataframe/datasource/read_raydataset.py:57]:
```suggestion
    def _tile_partitioned(cls, op: 'DataFrameReadRayDataset'):
```

File: mars/dataframe/datasource/read_raydataset.py [hunk 5]
```diff
@@ -0,0 +1,115 @@
...
+        out_df = op.outputs[0]
+        shape = (np.nan, out_df.shape[1])
+        dtypes = out_df.dtypes
+        dataset = op.refs
+
+        chunk_index = 0
+        out_chunks = []
+        for piece in dataset:
```
Comment (chaokunyang) [inline mars/dataframe/datasource/read_raydataset.py:65]:
```suggestion
        for object_ref in dataset:
```

File: mars/dataframe/datasource/read_raydataset.py [hunk 6]
```diff
@@ -0,0 +1,115 @@
...
+            chunk_op = op.copy().reset_key()
+            chunk_op._refs = piece
```
Comment (chaokunyang) [inline mars/dataframe/datasource/read_raydataset.py:67]: type of `_refs` seems to be ray.ObjectRef but not `List[ObjectRef]`?

File: mars/dataframe/datasource/read_raydataset.py [hunk 7]
```diff
@@ -0,0 +1,115 @@
...
+    index_value = parse_index(pd.RangeIndex(-1))
+    columns_value = parse_index(dtypes.index, store_data=True)
+
+    op = DataFrameReadRayDataset(refs=refs, columns=columns,
+                                 incremental_index=incremental_index)
+    return op(index_value=index_value, columns_value=columns_value,
+              dtypes=dtypes)
+
+
+def _get_dtypes(t):
```
Comment (chaokunyang) [inline mars/dataframe/datasource/read_raydataset.py:114]: For ray dataset, the ray object is an arrow.Table, does it has `dtypes`? DO we neec to read arrow table schema? Or just use the schema in ray.data.datset.DataSet

**Review by chaokunyang — APPROVED** [review]
LGTM

**Review by qinxuye — APPROVED** [review]
LGTM

**Review by wjsi — APPROVED** [review]
LGTM

**Pull request merged**